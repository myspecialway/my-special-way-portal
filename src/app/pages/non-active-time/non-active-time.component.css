/**
 * Flex-based table structure
 */
mat-table {
  display: block;
  direction: rtl;
  font-family: 'Rubik', 'Helvetica', 'Arial', sans-serif;
}

mat-paginator {
  direction: rtl;
}

mat-header-row {
  background-color: #f2f2f2;
  min-height: 48px;
}

mat-row,
mat-footer-row {
  min-height: 50px;
}

mat-row,
mat-header-row,
mat-footer-row {
  display: flex;
  border-width: 0;
  border-bottom-width: 1px;
  border-style: solid;
  align-items: center;
  box-sizing: border-box;
}
mat-row::after,
mat-header-row::after,
mat-footer-row::after {
  display: inline-block;
  min-height: inherit;
  content: '';
}

mat-cell:first-child,
mat-header-cell:first-child,
mat-footer-cell:first-child {
  padding-left: 24px;
}
[dir='rtl'] mat-cell:first-child,
[dir='rtl'] mat-header-cell:first-child,
[dir='rtl'] mat-footer-cell:first-child {
  padding-left: 0;
  padding-right: 24px;
}

mat-cell:last-child,
mat-header-cell:last-child,
mat-footer-cell:last-child {
  padding-right: 24px;
}
[dir='rtl'] mat-cell:last-child,
[dir='rtl'] mat-header-cell:last-child,
[dir='rtl'] mat-footer-cell:last-child {
  padding-right: 0;
  padding-left: 24px;
}

mat-cell,
mat-header-cell,
mat-footer-cell {
  flex: 1;
  display: flex;
  align-items: center;
  overflow: hidden;
  word-wrap: break-word;
  min-height: inherit;
}

/**
 * Native HTML table structure
 */
table.mat-table {
  border-spacing: 0;
}

tr.mat-header-row {
  height: 48px;
}

tr.mat-row,
tr.mat-footer-row {
  height: 50px;
}

th.mat-header-cell {
  text-align: left;
}

th.mat-header-cell,
td.mat-cell,
td.mat-footer-cell {
  padding: 0;
  border-bottom: 1px solid;
}

th.mat-header-cell:first-child,
td.mat-cell:first-child,
td.mat-footer-cell:first-child {
  padding-left: 24px;
}

th.mat-header-cell:last-child,
td.mat-cell:last-child,
td.mat-footer-cell:last-child {
  padding-right: 24px;
}

.msw-table-tools-header {
  align-items: center;
  display: flex;
  justify-content: space-between;
}

mat-row:hover {
  background-color: #e1ccf5;
  cursor: pointer;
}
mat-row:hover mat-cell {
  color: #8c37cd;
}
mat-row:hover .delete-row-button {
  display: block;
}

.delete-row-button {
  display: none;
}

.mat-raised-button.mat-primary {
  background-color: #6a00cf;
  border-radius: 3px;
}

.mat-tab-group.mat-primary .mat-ink-bar,
.mat-tab-nav-bar.mat-primary .mat-ink-bar {
  background-color: #ff0;
}

.header-table-count {
  width: 164px;
  height: 38px;
  font-size: 32px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: 0.6px;
  text-align: right;
  color: #222222;
}

.header-table-count-number {
  font-size: 18px;
}

.header-table-count-label {
  padding-left: 10px;
}

.header-menu-container {
  padding-bottom: 29px;
}

.msw-table-container .mat-cell:last-child,
.msw-table-container .mat-header-cell:last-child {
  flex-grow: 0;
  flex-basis: 100px;
}

.msw-table-container .mat-header-cell {
  font-weight: bold;
}

.no-records-msg {
  padding: 100px;
  text-align: center;
}

.filter-toggle-button {
  font-size: large;
  padding-top: 0.5rem;
}

.filter {
  max-width: 70%;
}

.material-icons.md-18 {
  font-size: 18px;
}

.material-icons.md-24 {
  font-size: 24px;
}

.material-icons.md-36 {
  font-size: 36px;
}

.material-icons.md-48 {
  font-size: 48px;
}

.header-table-count-label {
  font-size: 21px;
}

/*# sourceMappingURL=non-active-time.component.css.map */
