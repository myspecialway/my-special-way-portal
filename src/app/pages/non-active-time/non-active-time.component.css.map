{"version":3,"sourceRoot":"","sources":["../../../assets/scss/msw-table.scss","non-active-time.component.scss"],"names":[],"mappings":"AAIA;AAAA;AAAA;AAGA;EACE;EACA;EACA;;;AAGF;EACE;;;AAGF;EACE;EACA,YAnBsB;;;AAsBxB;AAAA;EAEE,YAvBe;;;AA0BjB;AAAA;AAAA;EAGE;EAGA;EACA;EACA;EACA;EACA;;AAKA;AAAA;AAAA;EACE;EACA;EACA;;;AAIJ;AAAA;AAAA;EAGE,cAlD2B;;AAoD3B;AAAA;AAAA;EACE;EACA,eAtDyB;;;AA0D7B;AAAA;AAAA;EAGE,eA7D2B;;AA+D3B;AAAA;AAAA;EACE;EACA,cAjEyB;;;AAqE7B;AAAA;AAAA;EAGE;EACA;EACA;EACA;EACA;EACA;;;AAGF;AAAA;AAAA;AAGA;EACE;;;AAGF;EACE,QA1FsB;;;AA6FxB;AAAA;EAEE,QA9Fe;;;AAiGjB;EACE;;;AAGF;AAAA;AAAA;EAGE;EACA;;;AAGF;AAAA;AAAA;EAGE,cA9G2B;;;AAiH7B;AAAA;AAAA;EAGE,eApH2B;;;AAuH7B;EACE;EACA;EACA;;;AAGF;EACE;EACA;;AACA;EACE;;AAEF;EACE;;;AAIJ;EACE;;;AAGF;EACE;EACA;;;AAGF;AAAA;EAEE;;;AAGF;EACE;EACA;EACA;EACA;EACA;EACA;EACA;EACA;EACA;EACA;;;AAEF;EACE;;;AAGF;EACE;;;AAGF;EACE;;;AAGF;AAAA;EAEE;EACA;;;AAEF;EACE;;;AAGF;EACE;EACA;;;AAGF;EACE;EACA;;;AAGF;EACE;;;AAGF;EACE;;;AAEF;EACE;;;AAEF;EACE;;;AAEF;EACE;;;AC/MF;EACE","file":"non-active-time.component.css"}