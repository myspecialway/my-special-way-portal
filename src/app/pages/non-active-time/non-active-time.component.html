<div data-test-id="non-active-time-page" class="main-content msw-non-active-time-container">
  <div class="container-fluid">
    <div class="header-menu-container msw-table-tools-header">
      <div class="header-table-count">
        <div>
          <span class="header-table-count-label">זמני אי פעילות</span>
          <span class="header-table-count-number">({{dataSource.data.length}})</span>
        </div>
      </div>
      <div class="example-header msw-table-tools-header">
        <button data-test-id="add-new-non-active-time-button" mat-raised-button color="primary" routerLink="/student/_new_">
          <i class="material-icons">
            add
          </i>
          הוסף זמן אי פעילות
        </button>
      </div>
    </div>
    <div class="row msw-table-container">
      <div class="col-lg-12 col-md-12 col-sm-12">
        <mat-table #table [dataSource]="dataSource" class="msw-table" matSort matSortActive="title" matSortDisableClear
                   matSortDirection="asc" >

          <!-- Name Column -->
          <ng-container matColumnDef="title" >
            <mat-header-cell *matHeaderCellDef mat-sort-header>אירוע</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="editNewNonActiveTime(row)">
              <div>{{row.title}}</div>
            </mat-cell>
          </ng-container>

          <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
          <mat-row *matRowDef="let row; columns: displayedColumns;"></mat-row>
        </mat-table>
      </div>
    </div>
  </div>
</div>
