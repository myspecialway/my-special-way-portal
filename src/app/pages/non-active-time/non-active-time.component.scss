@import '../../../assets/scss/msw-table';
