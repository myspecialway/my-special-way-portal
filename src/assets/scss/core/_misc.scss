body {
  background-color: #fafafa;
  color: $black-color;
  font-weight: 300;
}

legend {
  border-bottom: 0;
}

.serif-font {
  font-family: $font-family-serif;
}

// Prevent highlight on mobile
* {
  -webkit-tap-highlight-color: rgba(255, 255, 255, 0);
  -webkit-tap-highlight-color: transparent;
  &:focus {
    outline: 0;
  }
}

a {
  color: $link-color;
  &:hover,
  &:focus {
    color: darken($link-color, 5%);
    text-decoration: none;
  }

  &.text-info {
    &:hover,
    &:focus {
      color: darken($brand-info, 5%);
    }
  }

  & .material-icons {
    vertical-align: middle;
  }
}

.form-check,
label {
  font-size: 14px;
  line-height: 1.42857;
  color: $checkboxes-text-color;
  font-weight: 400;
}

/*           Animations              */
.animation-transition-general {
  @include transition-all($general-transition-time, $transition-linear);
}

.animation-transition-slow {
  @include transition-all($slow-transition-time, $transition-linear);
}

.animation-transition-fast {
  @include transition-all($fast-transition-time, $transition-ease);
}

.caret,
.sidebar a {
  @include transition-all($fast-transition-time, $transition-ease-in);
}
#map {
  position: relative;
  z-index: 2;
  width: 100%;
  height: calc(100vh - 70px);
  margin-top: 70px;
}

@font-face {
  font-family: 'pass';
  font-style: normal;
  font-weight: 400;
  src: url(data:application/font-woff;charset=utf-8;base64,d09GRgABAAAAAATsAA8AAAAAB2QAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABWAAAABwAAAAcg9+z70dERUYAAAF0AAAAHAAAAB4AJwANT1MvMgAAAZAAAAA/AAAAYH7AkBhjbWFwAAAB0AAAAFkAAAFqZowMx2N2dCAAAAIsAAAABAAAAAQAIgKIZ2FzcAAAAjAAAAAIAAAACAAAABBnbHlmAAACOAAAALkAAAE0MwNYJ2hlYWQAAAL0AAAAMAAAADYPA2KgaGhlYQAAAyQAAAAeAAAAJAU+ATJobXR4AAADRAAAABwAAAAcCPoA6mxvY2EAAANgAAAAEAAAABAA5gFMbWF4cAAAA3AAAAAaAAAAIAAKAE9uYW1lAAADjAAAARYAAAIgB4hZ03Bvc3QAAASkAAAAPgAAAE5Ojr8ld2ViZgAABOQAAAAGAAAABuK7WtIAAAABAAAAANXulPUAAAAA1viLwQAAAADW+JM4eNpjYGRgYOABYjEgZmJgBEI2IGYB8xgAA+AANXjaY2BifMg4gYGVgYVBAwOeYEAFjMgcp8yiFAYHBl7VP8wx/94wpDDHMIoo2DP8B8kx2TLHACkFBkYA8/IL3QB42mNgYGBmgGAZBkYGEEgB8hjBfBYGDyDNx8DBwMTABmTxMigoKKmeV/3z/z9YJTKf8f/X/4/vP7pldosLag4SYATqhgkyMgEJJnQFECcMOGChndEAfOwRuAAAAAAiAogAAQAB//8AD3jaY2BiUGJgYDRiWsXAzMDOoLeRkUHfZhM7C8Nbo41srHdsNjEzAZkMG5lBwqwg4U3sbIx/bDYxgsSNBRUF1Y0FlZUYBd6dOcO06m+YElMa0DiGJIZUxjuM9xjkGRhU2djZlJXU1UDQ1MTcDASNjcTFQFBUBGjYEkkVMJCU4gcCKRTeHCk+fn4+KSllsJiUJEhMUgrMUQbZk8bgz/iA8SRR9qzAY087FjEYD2QPDDAzMFgyAwC39TCRAAAAeNpjYGRgYADid/fqneL5bb4yyLMwgMC1H90HIfRkCxDN+IBpFZDiYGAC8QBbSwuceNpjYGRgYI7594aBgcmOAQgYHzAwMqACdgBbWQN0AAABdgAiAAAAAAAAAAABFAAAAj4AYgI+AGYB9AAAAAAAKgAqACoAKgBeAJIAmnjaY2BkYGBgZ1BgYGIAAUYGBNADEQAFQQBaAAB42o2PwUrDQBCGvzVV9GAQDx485exBY1CU3PQgVgIFI9prlVqDwcZNC/oSPoKP4HNUfQLfxYN/NytCe5GwO9/88+/MBAh5I8C0VoAtnYYNa8oaXpAn9RxIP/XcIqLreZENnjwvyfPieVVdXj2H7DHxPJH/2/M7sVn3/MGyOfb8SWjOGv4K2DRdctpkmtqhos+D6ISh4kiUUXDj1Fr3Bc/Oc0vPqec6A8aUyu1cdTaPZvyXyqz6Fm5axC7bxHOv/r/dnbSRXCk7+mpVrOqVtFqdp3NKxaHUgeod9cm40rtrzfrt2OyQa8fppCO9tk7d1x0rpiQcuDuRkjjtkHt16ctbuf/radZY52/PnEcphXpZOcofiEZNcQAAeNpjYGIAg///GBgZsAF2BgZGJkZmBmaGdkYWRla29JzKggxD9tK8TAMDAxc2D0MLU2NjENfI1M0ZACUXCrsAAAABWtLiugAA)
    format('woff');
}

input[type='password'] {
  font-family: 'pass', 'Roboto', Helvetica, Arial, sans-serif;
  font-size: 18px;
  &::-webkit-input-placeholder {
    transform: scale(0.77);
    transform-origin: 0 50%;
  }
  &::-moz-placeholder {
    font-size: 14px;
    opacity: 1;
  }
  &:-ms-input-placeholder {
    font-size: 14px;
    font-family: 'Roboto', Helvetica, Arial, sans-serif;
  }
}
