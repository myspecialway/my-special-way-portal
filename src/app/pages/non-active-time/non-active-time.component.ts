import { Component, OnInit, ViewChild } from '@angular/core';
import { MatDialog, MatSort, MatTableDataSource } from '@angular/material';
import { SubscriptionCleaner } from '../../decorators/SubscriptionCleaner.decorator';
import { Subscription } from 'rxjs';
import { NonActiveTimeService } from '../../services/non-active-time/non-active-time.graphql.service';
import { NonActiveTime } from '../../models/non-active-time.model';
import { first } from 'rxjs/operators';
import { DeleteNonActiveTimeDialogueComponent } from './delete/delete-non-active-time-dialogue.component';

@Component({
  selector: 'app-non-active-time',
  templateUrl: './non-active-time.component.html',
  styleUrls: ['./non-active-time.component.scss'],
})
export class NonActiveTimeComponent implements OnInit {
  displayedColumns = ['title', 'classes', 'dates', 'hours', 'deleteNonActiveTime'];
  dataSource = new MatTableDataSource<NonActiveTime>();

  @ViewChild(MatSort)
  sort: MatSort;

  @SubscriptionCleaner()
  subCollector: Subscription;

  constructor(private nonActiveTimeService: NonActiveTimeService, public dialog: MatDialog) {}

  async ngOnInit() {
    try {
      this.subCollector.add(
        this.nonActiveTimeService.getAllNonActiveTimes().subscribe((nonActiveTimes) => {
          if (nonActiveTimes !== null) {
            this.dataSource.data = [...nonActiveTimes];
          } else {
            this.dataSource.data = [];
          }
        }),
      );
      this.dataSource.sort = this.sort;
    } catch (error) {
      console.error('Error getting non active times');
      throw error;
    }
  }

  public async deleteNonActiveTime(nonActiveTime: NonActiveTime) {
    const dialogRef = this.dialog.open(DeleteNonActiveTimeDialogueComponent, {
      data: { _id: nonActiveTime._id, title: nonActiveTime.title },
    });
    this.subCollector.add(
      dialogRef
        .afterClosed()
        .pipe(first())
        .subscribe(async (result) => {
          if (result === true) {
            this.nonActiveTimeService.delete(nonActiveTime._id);
          }
        }),
    );
  }

  getClassesDisplayData(row: NonActiveTime): string {
    if (row.isAllClassesEvent) return 'כל הכיתות';
    if (!row.classes) throw new Error('Non active time is marked for specific classes, but no classes were specified');
    if (row.classes.length === 1) return row.classes[0].name;
    return row.classes[0].name + ' + ' + (row.classes.length - 1);
  }

  getDatesDisplayData(row: NonActiveTime): string {
    const startDateTime = new Date(row.startDateTime);
    const options = { year: 'numeric', month: 'long', day: 'numeric' };
    let result: string = startDateTime.toLocaleDateString('he-IL', options);
    if (row.startDateTime !== row.endDateTime) {
      const endDateTime = new Date(row.startDateTime);
      result += ' עד ' + endDateTime.toLocaleDateString('he-IL', options);
    }

    return result;
  }

  getHoursDisplayData(row: NonActiveTime): string {
    if (row.isAllDayEvent) return 'כל היום';
    const startDateTime = new Date(row.startDateTime);
    const endDateTime = new Date(row.endDateTime);
    const options = { hour: 'numeric', minute: 'numeric' };
    return (
      startDateTime.toLocaleTimeString('he-IL', options) + ' עד ' + endDateTime.toLocaleTimeString('he-IL', options)
    );
  }
}
