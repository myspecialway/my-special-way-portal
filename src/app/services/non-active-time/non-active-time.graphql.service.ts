import { Injectable } from '@angular/core';
import gql from 'graphql-tag';
import { Apollo } from 'apollo-angular';
import { Observable, of as observableOf, Subscription } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { NonActiveTime } from '../../models/non-active-time.model';
import { FetchResult } from 'apollo-link';
import { SubscriptionCleaner } from '../../decorators/SubscriptionCleaner.decorator';

const GET_ALL_NON_ACTIVE_TIMES_QUERY = gql`
  {
    nonActiveTimes {
      _id
      title
      isAllDayEvent
      startDateTime
      endDateTime
      isAllClassesEvent
      classes {
        _id
        name
      }
    }
  }
`;

const DELETE_NON_ACTIVE_TIME_QUERY = (id: string) => gql`
mutation {
  deleteNonActiveTime(
    id: "${id}"
  )
}`;

const CREATE_NON_ACTIVE_TIME_QUERY = (
  title: string,
  isAllDayEvent: boolean,
  startDateTime: string,
  endDateTime: string,
  isAllClassesEvent: boolean,
  classesIds: string[] | undefined,
) => gql`
  mutation {
    createNonActiveTime(nonActiveTime: {
      title: "${title}"
      isAllDayEvent: ${isAllDayEvent}
      startDateTime: "${startDateTime}"
      endDateTime: "${endDateTime}"
      isAllClassesEvent: ${isAllClassesEvent}
      classesIds: ${classesIds ? '[' + classesIds.map((classesId) => `"${classesId}"`) + ']' : undefined}
    }){ _id }
  }`;

const UPDATE_NON_ACTIVE_TIME_QUERY = (
  id: string,
  title: string,
  isAllDayEvent: boolean,
  startDateTime: string,
  endDateTime: string,
  isAllClassesEvent: boolean,
  classesIds: string[] | undefined,
) => gql`
 mutation {
  updateNonActiveTime(id: "${id}", nonActiveTime: {
    title: "${title}"
    isAllDayEvent: ${isAllDayEvent}
    startDateTime: "${startDateTime}"
    endDateTime: "${endDateTime}"
    isAllClassesEvent: ${isAllClassesEvent}
    classesIds: ${classesIds ? '[' + classesIds.map((classesId) => `"${classesId}"`) + ']' : undefined}
  }){ _id }
}`;

@Injectable()
export class NonActiveTimeService {
  @SubscriptionCleaner()
  subCollector: Subscription;

  constructor(private apollo: Apollo) {}

  public getAllNonActiveTimes(): Observable<NonActiveTime[]> {
    return this.apollo
      .watchQuery<{ nonActiveTimes: NonActiveTime[] }>({
        query: GET_ALL_NON_ACTIVE_TIMES_QUERY,
      })
      .valueChanges.pipe(
        map((res) => {
          return res.data.nonActiveTimes;
        }),
        catchError((err: TypeError) => {
          console.warn('user.component::ngInInit:: empty stream received');
          return observableOf([]);
        }),
      );
  }

  public create(
    title: string,
    isAllDayEvent: boolean,
    startDateTime: string,
    endDateTime: string,
    isAllClassesEvent: boolean,
    classesIds: string[] | undefined,
  ): any {
    return this.apollo
      .mutate({
        mutation: CREATE_NON_ACTIVE_TIME_QUERY(
          title,
          isAllDayEvent,
          startDateTime,
          endDateTime,
          isAllClassesEvent,
          classesIds,
        ),
        refetchQueries: [
          {
            query: GET_ALL_NON_ACTIVE_TIMES_QUERY,
          },
        ],
        awaitRefetchQueries: true,
      })
      .toPromise();
  }

  public update(
    id: string,
    title: string,
    isAllDayEvent: boolean,
    startDateTime: string,
    endDateTime: string,
    isAllClassesEvent: boolean,
    classesIds: string[] | undefined,
  ): any {
    return this.apollo
      .mutate({
        mutation: UPDATE_NON_ACTIVE_TIME_QUERY(
          id,
          title,
          isAllDayEvent,
          startDateTime,
          endDateTime,
          isAllClassesEvent,
          classesIds,
        ),
        refetchQueries: [
          {
            query: GET_ALL_NON_ACTIVE_TIMES_QUERY,
          },
        ],
        awaitRefetchQueries: true,
      })
      .toPromise();
  }

  public async delete(nonActiveTimeId: string): Promise<FetchResult<Record<string, any>>> {
    return this.apollo
      .mutate({
        mutation: DELETE_NON_ACTIVE_TIME_QUERY(nonActiveTimeId),
        refetchQueries: [
          {
            query: GET_ALL_NON_ACTIVE_TIMES_QUERY,
          },
        ],
        awaitRefetchQueries: true,
      })
      .toPromise();
  }
}
