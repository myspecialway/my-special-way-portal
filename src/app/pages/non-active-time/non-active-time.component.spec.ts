import { TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { MatHeaderRow, MatRowDef, MatHeaderRowDef, MatSort, MatPaginator, MatDialog } from '@angular/material';

import { Observable } from 'rxjs-compat';
import { NonActiveTimeService } from '../../services/non-active-time/non-active-time.graphql.service';
import { NonActiveTime } from '../../models/non-active-time.model';
import { NonActiveTimeComponent } from './non-active-time.component';

describe('non active time component', () => {
  let nonActiveTimeServiceMock: Partial<NonActiveTimeService>;
  let nonActiveTimeTestData: NonActiveTime[] = [];

  let nonActiveTimeDialogueMock: Partial<MatDialog>;
  let nonActiveTimeDialogueMockValue: boolean;

  beforeAll(async () => {
    nonActiveTimeTestData = [
      {
        _id: '5c0fa61642d6052794ab44c8',
        title: 'best title ever',
        isAllDayEvent: false,
        startDateTime: 'Mon, 10 Dec 2018 14:23:09 GMT',
        endDateTime: 'Tue, 11 Dec 2018 14:23:09 GMT',
        isAllClassesEvent: false,
        classes: [
          {
            _id: '5c04f3814408e8334404ef16',
            name: 'טיטאן',
          },
          {
            _id: '5c04f3814408e8334404ef17',
            name: 'פטל',
          },
        ],
      },
      {
        _id: '5c0fbbb04e773c17106afdb0',
        title: 'best title ever 2222222',
        isAllDayEvent: true,
        startDateTime: 'Mon, 10 Dec 2018 14:23:09 GMT',
        endDateTime: 'Mon, 10 Dec 2018 14:23:09 GMT',
        isAllClassesEvent: false,
        classes: [
          {
            _id: '5c04f3814408e8334404ef17',
            name: 'פטל',
          },
        ],
      },
      {
        _id: '5c0fbbb04e773c17106afdb0',
        title: 'best title ever 2222222',
        isAllDayEvent: true,
        startDateTime: 'Mon, 10 Dec 2018 14:23:09 GMT',
        endDateTime: 'Mon, 10 Dec 2018 14:23:09 GMT',
        isAllClassesEvent: true,
      },
    ];
  });

  beforeEach(async () => {
    nonActiveTimeServiceMock = {
      getAllNonActiveTimes: jest.fn(),
      delete: jest.fn(),
      create: jest.fn(),
      update: jest.fn(),
    };
    nonActiveTimeDialogueMockValue = true;
    nonActiveTimeDialogueMock = {
      open: jest.fn().mockReturnValue({
        afterClosed: jest.fn().mockImplementation(() => Observable.of(nonActiveTimeDialogueMockValue)),
      }),
    };

    TestBed.configureTestingModule({
      imports: [],
      declarations: [NonActiveTimeComponent, MatHeaderRow, MatRowDef, MatHeaderRowDef, MatSort, MatPaginator],
      providers: [
        { provide: NonActiveTimeService, useValue: nonActiveTimeServiceMock },
        { provide: MatDialog, useValue: nonActiveTimeDialogueMock },
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });
  });

  it('should render component as described in snapshot', () => {
    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    expect(fixture).toMatchSnapshot();
  });

  it('should load correct number of Non active times ', async () => {
    (nonActiveTimeServiceMock.getAllNonActiveTimes as jest.Mock).mockImplementationOnce(() => {
      return Observable.of(nonActiveTimeTestData);
    });
    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    await fixture.componentInstance.ngOnInit(); // this triggers the subCleaner instantiator.
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.dataSource.data.length).toEqual(3);
  });

  it('should load zero Non active times in case of error', async () => {
    (nonActiveTimeServiceMock.getAllNonActiveTimes as jest.Mock).mockImplementationOnce(() => {
      return Observable.throwError('testing with errors');
    });
    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    await fixture.componentInstance.ngOnInit(); // this triggers the subCleaner instantiator.
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.dataSource.data.length).toEqual(0);
  });

  it('should open delete component dialogue', async () => {
    (nonActiveTimeServiceMock.getAllNonActiveTimes as jest.Mock).mockImplementationOnce(() => {
      return Observable.of(nonActiveTimeTestData);
    });
    (nonActiveTimeServiceMock.delete as jest.Mock).mockImplementationOnce(() => {
      return Promise.resolve(1);
    });
    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    await fixture.componentInstance.ngOnInit(); // this triggers the subCleaner instantiator.
    await fixture.componentInstance.deleteNonActiveTime(nonActiveTimeTestData[0]);
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(nonActiveTimeDialogueMock.open).toHaveBeenCalled();
  });
  it('should call deleteNonActiveTime when user confirms ', async () => {
    (nonActiveTimeServiceMock.getAllNonActiveTimes as jest.Mock).mockImplementationOnce(() => {
      return Observable.of(nonActiveTimeTestData);
    });
    (nonActiveTimeServiceMock.delete as jest.Mock).mockImplementationOnce(() => {
      return Promise.resolve(1);
    });
    nonActiveTimeDialogueMockValue = true;
    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    await fixture.componentInstance.ngOnInit(); // this triggers the subCleaner instantiator.
    await fixture.componentInstance.deleteNonActiveTime(nonActiveTimeTestData[0]);
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(nonActiveTimeServiceMock.delete).toHaveBeenCalledWith(nonActiveTimeTestData[0]._id);
  });

  it('should not call deleteNonActiveTime when user cancels ', async () => {
    (nonActiveTimeServiceMock.getAllNonActiveTimes as jest.Mock).mockImplementationOnce(() => {
      return Observable.of(nonActiveTimeTestData);
    });
    (nonActiveTimeServiceMock.delete as jest.Mock).mockImplementationOnce(() => {
      return Promise.resolve(1);
    });
    nonActiveTimeDialogueMockValue = false;

    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    await fixture.componentInstance.ngOnInit(); // this triggers the subCleaner instantiator.
    await fixture.componentInstance.deleteNonActiveTime(nonActiveTimeTestData[0]);
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(nonActiveTimeServiceMock.delete).not.toHaveBeenCalled();
  });

  it('should get empty array if no Non active times are defined', async () => {
    (nonActiveTimeServiceMock.getAllNonActiveTimes as jest.Mock).mockImplementationOnce(() => {
      return Observable.of(null);
    });
    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    await fixture.componentInstance.ngOnInit(); // this triggers the subCleaner instantiator.
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.dataSource.data).toEqual([]);
  });

  it('should get correct class-display string for an all day event', async () => {
    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    await fixture.componentInstance.ngOnInit(); // this triggers the subCleaner instantiator.
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.getClassesDisplayData(nonActiveTimeTestData[2])).toEqual('כל הכיתות');
  });

  it('should get error for a non all day event with no classes', async () => {
    const badNonActiveTime: NonActiveTime = {
      _id: '5c0fbbb04e773c17106afdb0',
      title: 'best title ever 2222222',
      isAllDayEvent: true,
      startDateTime: 'Mon, 10 Dec 2018 14:23:09 GMT',
      endDateTime: 'Mon, 10 Dec 2018 14:23:09 GMT',
      isAllClassesEvent: false,
    };
    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    let err = false;
    try {
      fixture.componentInstance.getClassesDisplayData(badNonActiveTime);
    } catch (e) {
      err = true;
    }
    expect(err).toBeTruthy();
  });

  it('should get correct class-display string for an all day event with one class', async () => {
    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    await fixture.componentInstance.ngOnInit(); // this triggers the subCleaner instantiator.
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.getClassesDisplayData(nonActiveTimeTestData[1])).toEqual('פטל');
  });

  it('should get correct class-display string for an all day event with multiple classes', async () => {
    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    await fixture.componentInstance.ngOnInit(); // this triggers the subCleaner instantiator.
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.getClassesDisplayData(nonActiveTimeTestData[0])).toEqual('טיטאן + 1');
  });

  it('should get correct date-display string for an event startDate===endDate', async () => {
    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    await fixture.componentInstance.ngOnInit(); // this triggers the subCleaner instantiator.
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.getDatesDisplayData(nonActiveTimeTestData[1])).toEqual('2018 M12 10');
  });

  it('should get correct date-display string for an event startDate!==endDate', async () => {
    const fixture = TestBed.createComponent(NonActiveTimeComponent);
    await fixture.componentInstance.ngOnInit(); // this triggers the subCleaner instantiator.
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.getDatesDisplayData(nonActiveTimeTestData[0])).toEqual(
      '2018 M12 10 עד 2018 M12 10',
    );
  });
});
