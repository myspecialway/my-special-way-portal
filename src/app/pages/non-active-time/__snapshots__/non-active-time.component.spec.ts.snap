// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`non active time component should render component as described in snapshot 1`] = `
ZoneAwarePromise {
  "__zone_symbol__state": null,
  "__zone_symbol__value": Array [],
}
`;
